### PR TITLE
Use QPC instead of interlocked integer for epoch tracking

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Draft PR to use QPC to track the current epoch instead of an interlocked increment on a global variable.

Micro-benchmarks seem to show marked improvement in performance. Submitting a draft PR to get coverage from CI/CD pipeline.

### Baseline
```
test_epoch_enter_exit_alloc_free,1,1093
test_epoch_enter_exit_alloc_free,0,1003
```

### PR
```
test_epoch_enter_exit_alloc_free,1,565
test_epoch_enter_exit_alloc_free,0,372
```

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

**Comment by nibanks** [comment]
I have no context here, but have you considered per-CPU global variable instead of QPC. QPC can be very expensive. It regularly shows up in MsQuic performance traces as a non-trivial CPU usage. If each CPU epoch is independent, a per-CPU counter seems like it might work even better.

File: libs/platform/ebpf_epoch.c [hunk 1]
```diff
@@ -232,10 +228,11 @@ ebpf_epoch_exit()
     }
 
     if (ebpf_is_preemptible()) {
-        _ebpf_epoch_update_thread_state(current_cpu, ebpf_get_current_thread_id(), _ebpf_current_epoch, false);
+        _ebpf_epoch_update_thread_state(
+            current_cpu, ebpf_get_current_thread_id(), ebpf_query_time_since_boot(true), false);
     } else {
         ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
-        _ebpf_epoch_cpu_table[current_cpu].cpu_epoch_state.epoch = _ebpf_current_epoch;
+        _ebpf_epoch_cpu_table[current_cpu].cpu_epoch_state.epoch = ebpf_query_time_since_boot(true);
```
Comment (Alan-Jowett) [inline libs/platform/ebpf_epoch.c:235]: Do we actually need to capture time at the epoch exit?

**Comment by Alan-Jowett** [comment]
@nibanks thanks for the suggestion, I appreciate it. Going to create an discussion page on this, will tag you in it.

Unfortunately, we need to reason about events that occur across CPUs. Specifically, this counter is being used as part of the memory reclamation strategy for [RCU](https://en.wikipedia.org/wiki/Read-copy-update).

The code currently maintains two global counters:
1) [Current Epoch](https://github.com/microsoft/ebpf-for-windows/blob/d8f00f851382ba5b3ef574a479ab70aa80bf05c8/libs/platform/ebpf_epoch.c#L71) (CE)
2) [Release Epoch](https://github.com/microsoft/ebpf-for-windows/blob/d8f00f851382ba5b3ef574a479ab70aa80bf05c8/libs/platform/ebpf_epoch.c#L76) (RE)

The sequence of events is:
1) At the [start of each RCU region](https://github.com/microsoft/ebpf-for-windows/blob/d8f00f851382ba5b3ef574a479ab70aa80bf05c8/libs/platform/ebpf_epoch.c#L207), we capture the CE (unsynchronized read) and mark the RCU region as active
2) When [freeing memory](https://github.com/microsoft/ebpf-for-windows/blob/d8f00f851382ba5b3ef574a479ab70aa80bf05c8/libs/platform/ebpf_epoch.c#L275) the code does the following:
a) Capture CE (read via interlocked increment)
b) Tag the memory with the CE 
c) Insert it into a per-CPU free list
3) At the [end of the RCU region](https://github.com/microsoft/ebpf-for-windows/blob/d8f00f851382ba5b3ef574a479ab70aa80bf05c8/libs/platform/ebpf_epoch.c#L227), we then free any memory with an epoch older than Release Epoch
4) Mark the RCU region as inactive

A [period timer](https://github.com/microsoft/ebpf-for-windows/blob/d8f00f851382ba5b3ef574a479ab70aa80bf05c8/libs/platform/ebpf_epoch.c#L250) then computes a new RE base on the minimum CE and all active RCU region's copies of CE.

Taking a look at the profile showed that the interlocked increment on CE and the unsynchronized read of CE are both expensive due to cache line contention.

My original thought was to use the [TSC](https://en.wikipedia.org/wiki/Time_Stamp_Counter) as CE (instead of a global variable). The problem is that TSC isn't perfectly synchronized across CPUs (I have observed it being up to a few hundred cycles off after HAL synchronizes it). In addition, when a VM live migrates between CPU's TSC can roll back.

The platform offers QPC as a sanitized version of TSC with the following changes:
1) TSC divided by 1024 to prevent rollback between CPUs.
2) Hyper-V's reference count is added (computed so that TSC in the guest starts at zero). 

Preliminary results, with the baseline cut short due to an issue on the test machine. Tests were executed 50 times for QPC change and 10 times for baseline (baseline was interrupted). 

Data seems to indicate that the change raises the cost for RCU enter by around 50ns (I think we can do better) but reduces an RCU operation by around 500ns. Trivial operations are test_epoch_enter_exit (enter/leave RCU region) and test_epoch_enter_exit_alloc_free (enter region, perform RCU, exit region).

Test | Time (QPC) | STDEV (QPC) | Time (Base) | STDEV (Base) | Delta
-- | -- | -- | -- | -- | --
test_bpf_get_prandom_u32 | 2233 | 11 | 2187 | 10 | -2%
test_bpf_get_smp_processor_id | 123 | 7 | 77 | 6 | -37%
test_bpf_ktime_get_boot_ns | 140 | 5 | 91 | 6 | -35%
test_bpf_ktime_get_ns | 142 | 6 | 92 | 7 | -35%
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_ARRAY> | 129 | 7 | 82 | 7 | -36%
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_HASH> | 173 | 8 | 120 | 7 | -31%
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_LRU_HASH> | 1727 | 181 | 2242 | 218 | 30%
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_PERCPU_ARRAY> | 140 | 7 | 93 | 7 | -33%
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_PERCPU_HASH> | 185 | 6 | 131 | 6 | -29%
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_ARRAY> | 357 | 49 | 313 | 55 | -12%
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_HASH> | 204 | 31 | 142 | 13 | -30%
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_LRU_HASH> | 1695 | 254 | 2155 | 83 | 27%
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_PERCPU_ARRAY> | 143 | 8 | 93 | 6 | -35%
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_PERCPU_HASH> | 194 | 23 | 135 | 6 | -30%
test_bpf_map_update_elem<BPF_MAP_TYPE_ARRAY> | 240 | 39 | 209 | 35 | -13%
test_bpf_map_update_elem<BPF_MAP_TYPE_HASH> | 1663 | 235 | 2128 | 106 | 28%
test_bpf_map_update_elem<BPF_MAP_TYPE_LRU_HASH> | 2425 | 201 | 3682 | 75 | 52%
test_bpf_map_update_elem<BPF_MAP_TYPE_PERCPU_ARRAY> | 153 | 9 | 97 | 7 | -37%
test_bpf_map_update_elem<BPF_MAP_TYPE_PERCPU_HASH> | 204 | 8 | 143 | 4 | -30%
test_bpf_map_update_lru_elem<BPF_MAP_TYPE_LRU_HASH> | 4320528 | 56785 | 4431714 | 77522 | 3%
test_ebpf_hash_table_find | 144 | 2 | 86 | 3 | -40%
test_ebpf_hash_table_next_key | 158 | 3 | 98 | 4 | -38%
test_ebpf_hash_table_update | 1097 | 62 | 1962 | 44 | 79%
test_ebpf_hash_table_update_overlapping | 1329 | 67 | 2155 | 29 | 62%
test_epoch_enter_exit | 104 | 2 | 57 | 4 | -45%
test_epoch_enter_exit_alloc_free | 368 | 4 | 994 | 21 | 170%
test_lpm_trie_ipv4<1024> | 2471 | 26 | 2407 | 16 | -3%
test_lpm_trie_ipv4<1048576> | 3350 | 18 | 3322 | 72 | -1%
test_lpm_trie_ipv4<16384> | 2499 | 11 | 2444 | 10 | -2%
test_lpm_trie_ipv4<262144> | 2754 | 26 | 2696 | 11 | -2%
test_program_invoke_interpret | 223 | 6 | 171 | 4 | -24%
test_program_invoke_jit | 207 | 5 | 150 | 3 | -28%

Notes:
BPF_MAP_TYPE_LRU_HASH needs work - #557
Tail call check make test_program_invoke_jit >> test_epoch_enter_exit - #453

**Comment by Alan-Jowett** [comment]
Looks like the cost of QPC offsets the gain from modifying the shared pages.

**Pull request closed without merging**